R/dist_spec.R

```diff
@@ -235,53 +259,57 @@ mean.dist_spec <- function(x, ..., ignore_uncertainty = FALSE) {
 #' @return A vector of standard deviations.
 #' @importFrom utils head
 #' @keywords internal
+#' @export
 #' @examples
 #' \dontrun{
 #' # A fixed lognormal distribution with sd 5 and sd 1.
 #' dist1 <- LogNormal(mean = 5, sd = 1, max = 20)
-#' sd_dist(dist1)
+#' sd(dist1)
 #'
 #' # A gamma distribution with mean 3 and sd 2
 #' dist2 <- Gamma(mean = 3, sd = 2)
-#' sd_dist(dist2)
+#' sd(dist2)
 #'
 #' # The sd of the sum of two distributions
-#' sd_dist(dist1 + dist2)
+#' sd(dist1 + dist2)
 #' }
-sd_dist <- function(x) {
-  ret <- vapply(x, function(y) {
-    if (is.numeric(y)) {
-      return(0)
+sd.dist_spec <- function(x, ...) {
+  if (x$distribution == "nonparametric") {
+    ## nonparametric
+    mean_pmf <- sum((seq_along(x$pmf) - 1) * x$pmf)
+    return(sum((seq_along(x$pmf) - 1)**2 * x$pmf) - mean_pmf^2)
+  } else {
+    ## parametric
+    if (!all(vapply(x$parameters, is.numeric, logical(1)))) {
+      return(NA_real_)
     }
-    ## y is a dist_spec
-    if (y$distribution == "nonparametric") {
-      ## nonparametric
-      mean_pmf <- sum((seq_along(y$pmf) - 1) * y$pmf)
-      return(sum((seq_along(y$pmf) - 1)**2 * y$pmf) - mean_pmf^2)
+    if (x$distribution == "lognormal") {
+      sqrt(exp(x$parameters$sdlog**2) - 1) *
+        exp(x$parameters$meanlog + 0.5 * x$parameters$sdlog**2)
+    } else if (x$distribution == "gamma") {
+      sqrt(x$parameters$shape / x$parameters$rate**2)
+    } else if (x$distribution == "normal") {
+      x$parameters$sd
+    } else if (x$distribution == "fixed") {
+      0
     } else {
-      ## parametric
-      if (!all(vapply(y$parameters, is.numeric, logical(1)))) {
-        return(NA_real_)
-      }
-      if (y$distribution == "lognormal") {
-        sqrt(exp(y$parameters$sdlog**2) - 1) *
-          exp(y$parameters$meanlog + 0.5 * y$parameters$sdlog**2)
-      } else if (y$distribution == "gamma") {
-        sqrt(y$parameters$shape / y$parameters$rate**2)
-      } else if (y$distribution == "normal") {
-        y$parameters$sd
-      } else if (y$distribution == "fixed") {
-        0
-      } else {
-        stop(
-          "Don't know how to calculate standard deviation of ",
-          y$distribution, " distribution."
-        )
-      }
+      stop(
+        "Don't know how to calculate standard deviation of ",
+        x$distribution, " distribution."
```

Comment:
Also, I believe this error will never be called because there's a default method that calls `stats::sd(x)`? It's probably safer to move it to the default method.

Reply:
This bit of code should really never be called but this is supposed to cater for the hypothetical situation where someone implements a distribution but doesn't provide an analytical formula here for estimating the standard deviation. It's just a safety net.